### PR TITLE
Implement the `PauliSentence` and `PauliWord` classes 

### PR DESCRIPTION
**Pull request by Jaybsoni**
**Context:**
The second step to better support Pauli arithmetic in PennyLane. The reduced representation classes (outlined [here](https://github.com/PennyLaneAI/adrs/pull/52)) are implemented. 

**Description of the Change:**
Two new classes introduced `PauliSentence` and `PauliWord` in a `pennylane/pauli` module. 

**Benefits:**
None so far, there is no integration between these classes and the rest of PennyLane in this PR. Ultimately these will be used to replace existing functions in PL to perform arithmetic with Pauli operators.

### CODE REVIEW

**Comment by Jaybsoni** [comment]
[sc-28236]

File: pennylane/math/multi_dispatch.py [hunk 1]
```diff
@@ -152,6 +153,14 @@ def wrapper(*args, **kwargs):
...
 
 
+@multi_dispatch(argnum=[0])
+def kron(*args, like=None, **kwargs):
+    """The kronecker/tensor product of args."""
+    if like == "scipy":
+        return onp.kron(*args, **kwargs)  # Dispatch scipy kron to numpy backed specifically.
+    return np.kron(*args, **kwargs)
+
+
```
Comment (Jaybsoni) [inline pennylane/math/multi_dispatch.py:163]: This was required to avoid a deprecation warning as `scipy.kron` will be deprecated in the next cycle !

File: pennylane/math/multi_dispatch.py [hunk 2]
```diff
@@ -754,8 +763,11 @@ def convert(val):
     return [convert(val) for val in values]
 
 
-def add(*args, **kwargs):
+@multi_dispatch(argnum=[0])
+def add(*args, like=None, **kwargs):
     """Add arguments element-wise."""
+    if like == "scipy":
+        return onp.add(*args, **kwargs)  # Dispatch scipy add to numpy backed specifically.
```
Comment (Jaybsoni) [inline pennylane/math/multi_dispatch.py:770]: Same as `kron` for `add`

File: pennylane/pauli/pauli_arithmetic.py [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+    >>> w = PauliWord({"a": X, 2: Y, 3: Z})
+    """
+
+    def __missing__(self, key):
+        """If the wire is not in the Pauli word,
+        then no operator acts on it, so return the Identity."""
+        return I
+
+    def __init__(self, mapping={}):  # pylint:disable=dangerous-default-value
```
Comment (albi3ro) [inline pennylane/pauli/pauli_arithmetic.py:90]:
```suggestion
    def __init__(self, mapping):
```

How about just leave it as a mandatory argument?

File: pennylane/pauli/pauli_arithmetic.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+                    result[wire] = new_op
+            elif term != I:
+                result[wire] = term
+
+        return PauliWord(result), coeff
+
+    @property
+    def wires(self):
+        """Track wires in a PauliWord."""
+        return set(k for k in self.keys())
```
Comment (albi3ro) [inline pennylane/pauli/pauli_arithmetic.py:147]:
```suggestion
        return set(self)
```

Comment (albi3ro) [reply pennylane/pauli/pauli_arithmetic.py:147]: Also, do we really this property?

File: pennylane/pauli/pauli_arithmetic.py [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+
+            if len(pw) == 0:
+                rep_str += "[()]\n"
+                continue
+
+            for index_j, (w, op) in enumerate(pw.items()):
+                if index_j == 0:
+                    rep_str += f"[{op}({w})]"
+                else:
+                    rep_str += f" @ [{op}({w})]"
```
Comment (albi3ro) [inline pennylane/pauli/pauli_arithmetic.py:233]: Maybe define a `__str__` method for `PauliWord` too and just call that here?

Comment (albi3ro) [inline pennylane/pauli/pauli_arithmetic.py:233]: Also, using join is generally preferable to 
```
" @ ".join(f"{op}({w})" for w, op in self.items())
```

File: pennylane/pauli/pauli_arithmetic.py (hunk 3 above)
Comment (Jaybsoni) [reply pennylane/pauli/pauli_arithmetic.py]: Makes sense!

File: pennylane/pauli/pauli_arithmetic.py [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+        matrix_map = sparse_mat_map if format != "dense" else mat_map
+        kron = sparse.kron if format != "dense" else math.kron
+
+        initial_wire = wire_order[0]
+        result_matrix = matrix_map[self[initial_wire]]
+
+        for wire in wire_order[1:]:
+            result_matrix = kron(result_matrix, matrix_map[self[wire]])
+
+        return result_matrix
```
Comment (albi3ro) [inline pennylane/pauli/pauli_arithmetic.py:175]:
```suggestion
        return reduce(kron, (matrix_map[self[w]] for w in wire_order))
```
This is probably a good case for using `functools.reduce`. A bit more brevity.

File: pennylane/pauli/pauli_arithmetic.py [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+                matrix, then the format for the sparse representation of the matrix.
+
+        Returns:
+            (Union[NumpyArray, ScipySparseArray]): Matrix representation of the PauliSentence.
+
+        Rasies:
+            ValueError: Can't get the matrix of an empty PauliSentence.
+        """
+        if len(self) == 0:
+            raise ValueError("Can't get the matrix of an empty PauliSentence.")
```
Comment (albi3ro) [inline pennylane/pauli/pauli_arithmetic.py:258]: Identity?

File: pennylane/pauli/pauli_arithmetic.py [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+                matrix, then the format for the sparse representation of the matrix.
+
+        Returns:
+            (Union[NumpyArray, ScipySparseArray]): Matrix representation of the Pauliword
+
+        Raises:
+            ValueError: Can't get the matrix of an empty PauliWord.
+        """
+        if len(self) == 0:
+            raise ValueError("Can't get the matrix of an empty PauliWord.")
```
Comment (albi3ro) [inline pennylane/pauli/pauli_arithmetic.py:164]: Identity?

File: pennylane/pauli/pauli_arithmetic.py (hunk 7 above)
Comment (Jaybsoni) [reply pennylane/pauli/pauli_arithmetic.py:239]: An empty `PauliSentence` and `PauliWord` act kind of like the identity on all wires. So I would be hesitant to return a fixed dimension matrix here since the behaviours wouldn't match. I think having a error is fine 👍🏼

File: pennylane/pauli/pauli_arithmetic.py (hunk 8 above)
Comment (Jaybsoni) [reply pennylane/pauli/pauli_arithmetic.py:171]: An empty PauliSentence and PauliWord act kind of like the identity on all wires. So I would be hesitant to return a fixed dimension matrix here since the behaviours wouldn't match. I think having a error is fine 👍🏼

File: tests/math/test_multi_dispatch.py [hunk 9]
```diff
@@ -155,3 +156,15 @@ def test_unwrap():
...
+
+
+def test_kron():
+    """Test that a deprecation warning is avoided when arrays
+    are dispatched to scipy.kron."""
+    m1 = s.array([[0, 1], [1, 0]])
+    m2 = s.array([[1, 0], [0, 1]])
+    expected_result = onp.kron(m1, m2)
+
+    assert np.allclose(expected_result, fn.kron(m1, m2))
```
Comment (Jaybsoni) [inline tests/math/test_multi_dispatch.py:168]:
```suggestion
    assert np.allclose(expected_result, fn.kron(m1, m2, like="scipy"))
```

File: tests/pauli/test_pauli_arithmetic.py [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+
+        assert pw_1.__hash__() == pw_2.__hash__()
+        assert pw_1.__hash__() == pw_3.__hash__()
+        assert pw_1.__hash__() != pw_4.__hash__()
+
+    @pytest.mark.parametrize("pw", (pw1, pw2, pw3, pw4))
+    def test_copy(self, pw):
+        """Test that the copy is identical to the original."""
+        copy_pw = copy(pw)
+        assert copy_pw == pw
```
Comment (AlbertMitjans) [inline tests/pauli/test_pauli_arithmetic.py:84]: Check that is not the same instance
```suggestion
        assert copy_pw == pw
        assert copy_pw is not pw
```

File: pennylane/pauli/pauli_arithmetic.py [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+            elif term != I:
+                result[wire] = term
+
+        return PauliWord(result), coeff
+
+    def __str__(self):
+        """String representation of a PauliWord."""
+        if len(self) == 0:
+            return "[()]"
+        return " @ ".join(f"[{op}({w})]" for w, op in self.items())
```
Comment (AlbertMitjans) [inline pennylane/pauli/pauli_arithmetic.py:149]: Why did you add square brackets around each Pauli op? Why not:
```pycon
>>> print(PauliWord({0: "X", 1: "Y"}))
X(0) @ Y(1)
```

File: pennylane/pauli/pauli_arithmetic.py [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+                    del result[wire]
+                else:
+                    coeff *= factor
+                    result[wire] = new_op
+            elif term != I:
+                result[wire] = term
+
+        return PauliWord(result), coeff
+
+    def __str__(self):
```
Comment (AlbertMitjans) [inline pennylane/pauli/pauli_arithmetic.py:145]: Would be nice to add a `__repr__` method too

Comment (AlbertMitjans) [reply pennylane/pauli/pauli_arithmetic.py:145]: Right now:
```pycon
>>> PauliWord({0: "X", 1: "Y"})
{0: 'X', 1: 'Y'}
```

File: pennylane/pauli/pauli_arithmetic.py [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+        and the tensor product otherwise.
+
+        Args:
+            other (PauliWord): The Pauli word to multiply with
+
+        Returns:
+            result(PauliWord): The resulting operator of the multiplication
+            coeff(complex): The complex phase factor
+        """
+        c_self, c_other = (copy(self), copy(other))
```
Comment (AlbertMitjans) [inline pennylane/pauli/pauli_arithmetic.py:122]: Why do we copy both PWs? We are copying a lot of Pauli ops that will be modified/removed afterwards.

I believe it would be faster if we created an empty dict, populated it with the multiplication results and then casted it to a `PauliWord` instead of doing these 2 copies

File: pennylane/pauli/pauli_arithmetic.py [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+
+    def __missing__(self, key):
+        """If the PauliWord is not in the sentence then the coefficient
+        associated with it should be 0."""
+        return 0.0
+
+    def __add__(self, other):
+        """Add two Pauli sentence together by iterating over the smaller
+        one and adding its terms to the larger one."""
+        c_self, c_other = (copy(self), copy(other))
```
Comment (AlbertMitjans) [inline pennylane/pauli/pauli_arithmetic.py:197]: We just need to copy the `larger_ps`

File: pennylane/pauli/pauli_arithmetic.py [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+        """Multiply two Pauli sentences by iterating over each sentence and multiplying
+        the Pauli words pair-wise"""
+        final_ps = PauliSentence()
+
+        for pw1 in self:
+            for pw2 in other:
+                prod_pw, coeff = pw1 * pw2
+                final_ps[prod_pw] += coeff * self[pw1] * other[pw2]
+
+        return final_ps
```
Comment (AlbertMitjans) [inline pennylane/pauli/pauli_arithmetic.py:214]: Nice! We could do sth similar with the `PauliWord` class. No need to copy both classes.

File: tests/pauli/test_pauli_arithmetic.py [hunk 16]
```diff
@@ -0,0 +1,321 @@
...
+        (pw1, pw1, PauliWord({}), 1.0),  # identities are automatically removed !
+        (pw1, pw3, PauliWord({0: Z, 1: X, 2: Y, "b": Z, "c": Z}), 1.0),
+        (pw2, pw3, PauliWord({"a": X, "b": Y, 0: Z}), -1.0j),
+        (pw3, pw4, pw3, 1.0),
+    )
+
+    @pytest.mark.parametrize("pw1, pw2, result_pw, coeff", tup_pws_mult)
+    def test_mul(self, pw1, pw2, result_pw, coeff):
+        copy_pw1 = copy(pw1)
+        copy_pw2 = copy(pw2)
```
Comment (AlbertMitjans) [inline tests/pauli/test_pauli_arithmetic.py:114]: Why copying the classes? They should NOT be modified during the `__mult__` operation

File: pennylane/pauli/pauli_arithmetic.py [hunk 17]
```diff
@@ -0,0 +1,261 @@
...
+
+    def __str__(self):
+        """String representation of the PauliSentence."""
+        return "\n+ ".join(f"({coeff}) * {str(pw)}" for pw, coeff in self.items())
```
Comment (AlbertMitjans) [inline pennylane/pauli/pauli_arithmetic.py:218]: Instead of:
```pycon
>>> print(PauliSentence({PauliWord({0: "X", 1: "Y"}): 0.5j}))
(0.5j) * [X(0)] @ [Y(1)]
```
I would prefer:
```pycon
>>> print(PauliSentence({PauliWord({0: "X", 1: "Y"}): 0.5j}))
0.5j * (X(0) @ Y(1))
```

File: pennylane/pauli/pauli_arithmetic.py [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+        final_ps = PauliSentence()
+
+        for pw1 in self:
+            for pw2 in other:
+                prod_pw, coeff = pw1 * pw2
+                final_ps[prod_pw] += coeff * self[pw1] * other[pw2]
+
+        return final_ps
+
+    def __str__(self):
```
Comment (AlbertMitjans) [inline pennylane/pauli/pauli_arithmetic.py:216]: Add `__repr__` method

File: pennylane/pauli/pauli_arithmetic.py [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+
+    def __missing__(self, key):
+        """If the wire is not in the Pauli word,
+        then no operator acts on it, so return the Identity."""
+        return I
+
+    def __init__(self, mapping):
+        """Strip identities from PauliWord on init!"""
+        mapping = {wire: op for wire, op in mapping.items() if op != I}
+        super().__init__(mapping)
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/pauli/pauli_arithmetic.py:94]: The point of this class is performance, so I would avoid adding anything not strictly necessary.  We could always add checks like that in a higher-level wrapper if we want to expose it to users more.

File: pennylane/pauli/pauli_arithmetic.py [hunk 20]
```diff
@@ -0,0 +1,261 @@
...
+
+    def __copy__(self):
+        """Copy the PauliWord instance."""
+        return PauliWord(dict(self.items()))
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/pauli/pauli_arithmetic.py:98]: Double checked. Removing this removes the ability to copy.  Seems like the parent `__copy__` uses `__setitem__`, which we have set as not implemented.

File: pennylane/pauli/pauli_arithmetic.py [hunk 21]
```diff
@@ -0,0 +1,261 @@
...
+
+    def __setitem__(self, key, item):
+        """Restrict setting items after instantiation."""
+        raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/pauli/pauli_arithmetic.py:102]: Screen Shot 2022-10-31 at 11 25 52 AM

Maybe something like the error raised by trying to mutate an immutable object.

File: pennylane/pauli/pauli_arithmetic.py (hunk 8 above)
Comment (albi3ro) [reply pennylane/pauli/pauli_arithmetic.py:171]: Maybe `if len(self) == 0 and wire_order is None`?
If we have a `wire_order`, then we return an identity of the right size as wire order.

File: pennylane/pauli/pauli_arithmetic.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/pauli/pauli_arithmetic.py:98]: Yes exactly

File: pennylane/pauli/pauli_arithmetic.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/pauli/pauli_arithmetic.py:94]: I agree, I think we want this to be a more internal representation that we allow the user to interface with using helper/constructor functions

File: pennylane/pauli/pauli_arithmetic.py [hunk 22]
```diff
@@ -0,0 +1,261 @@
...
+        if len(self) == 0:
+            raise ValueError("Can't get the matrix of an empty PauliWord.")
+
+        matrix_map = sparse_mat_map if format != "dense" else mat_map
+        kron = sparse.kron if format != "dense" else math.kron
+
+        return reduce(kron, (matrix_map[self[w]] for w in wire_order))
+
+
+class PauliSentence(dict):
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/pauli/pauli_arithmetic.py:179]: Re: Christina's comment above

File: pennylane/pauli/pauli_arithmetic.py [hunk 23]
```diff
@@ -0,0 +1,261 @@
...
+            for pw, coeff in self.items()
+        )
+
+        reduced_mat, result_wire_order = math.reduce_matrices(
+            mats_and_wires_gen=mats_and_wires_gen, reduce_func=math.add
+        )
+
+        return math.expand_matrix(reduced_mat, result_wire_order, wire_order=wire_order)
+
+    def simplify(self, tol=1e-8):
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/pauli/pauli_arithmetic.py:256]: Just worried about having to copy large Hamiltonians here.

File: tests/math/test_multi_dispatch.py [hunk 24]
```diff
@@ -155,3 +156,15 @@ def test_unwrap():
...
+def test_kron():
+    """Test that a deprecation warning is avoided when arrays
+    are dispatched to scipy.kron."""
+    m1 = s.array([[0, 1], [1, 0]])
+    m2 = s.array([[1, 0], [0, 1]])
+    expected_result = onp.kron(m1, m2)
+
+    assert np.allclose(expected_result, fn.kron(m1, m2, like="scipy"))
+    with pytest.warns(DeprecationWarning):
+        _ = s.kron(m1, m2)
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply tests/math/test_multi_dispatch.py:170]: I don't think it matters? `s.kron` will eventually be removed completely!

File: pennylane/pauli/pauli_arithmetic.py [hunk 25]
```diff
@@ -0,0 +1,261 @@
...
+        """Remove any PauliWords in the PauliSentence with coefficients less than the threshold tolerance."""
+        terms = list(self.keys())
+        for pw in terms:
+            if abs(self[pw]) <= tol:
+                del self[pw]
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/pauli/pauli_arithmetic.py:261]: Nice clean up!

File: pennylane/pauli/pauli_arithmetic.py (hunk 11 above)
Comment (Jaybsoni) [reply pennylane/pauli/pauli_arithmetic.py:149]: Just a stylistic choice, updating now

File: tests/pauli/test_pauli_arithmetic.py (hunk 16 above)
Comment (Jaybsoni) [reply tests/pauli/test_pauli_arithmetic.py:117]: This was old, has been removed since

File: pennylane/pauli/pauli_arithmetic.py (hunk 17 above)
Comment (Jaybsoni) [reply pennylane/pauli/pauli_arithmetic.py]: Updated 👍🏼

File: pennylane/pauli/pauli_arithmetic.py (hunk 18 above)
Comment (Jaybsoni) [reply pennylane/pauli/pauli_arithmetic.py:219]: Added 👍🏼

File: pennylane/pauli/pauli_arithmetic.py (hunk 14 above)
Comment (Jaybsoni) [reply pennylane/pauli/pauli_arithmetic.py]: Nice catch 👍🏼

File: pennylane/pauli/pauli_arithmetic.py (hunk 13 above)
Comment (Jaybsoni) [reply pennylane/pauli/pauli_arithmetic.py]: The copies were added after the fact. Initially, the idea was that if we allow for one of the inputs to get mutated, then we can compute the multiplication in $~ O(min(m, n))$ time, where m, n are the number of factors in each Pauliword respectively. By adding the copy, the complexity becomes: $~ O(max(m, n))$. I was able to remove one of the two copies, but the time complexity remains the same. If we were to populate an empty dict, we would have to iterate over both PWs and keep track of terms that have already been multiplied. This would end up being the same time complexity $~ O(max(m, n))$. For this reason I think we should stick with this implementation.

File: pennylane/pauli/pauli_arithmetic.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/pauli/pauli_arithmetic.py]: Added 👍🏼

File: pennylane/pauli/pauli_arithmetic.py (hunk 15 above)
Comment (Jaybsoni) [reply pennylane/pauli/pauli_arithmetic.py:217]: See comment above

File: pennylane/pauli/pauli_arithmetic.py [hunk 26]
```diff
@@ -0,0 +1,261 @@
...
+    """
+
+    def __missing__(self, key):
+        """If the wire is not in the Pauli word,
+        then no operator acts on it, so return the Identity."""
+        return I
+
+    def __init__(self, mapping):
+        """Strip identities from PauliWord on init!"""
+        mapping = {wire: op for wire, op in mapping.items() if op != I}
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/pauli/pauli_arithmetic.py]: Added 👍🏼

File: pennylane/pauli/pauli_arithmetic.py [hunk 27]
```diff
@@ -0,0 +1,280 @@
...
+
+    @property
+    def wires(self):
+        """Track wires of the PauliSentence."""
+        return set().union(*(pw.wires for pw in self.keys()))
+
+    def to_mat(self, wire_order, format="dense"):
+        """Given a Pauli word, get the matrix representation.
+
+        KeywordArgs:
```
Comment (albi3ro) [inline pennylane/pauli/pauli_arithmetic.py:241]:
```suggestion
        Keyword Args:
```
Some kind of rendering issue in the documentation. Maybe this will fix it?

File: pennylane/pauli/pauli_arithmetic.py [hunk 28]
```diff
@@ -0,0 +1,280 @@
...
+        """Terminal representation for PauliSentence"""
+        return str(self)
+
+    @property
+    def wires(self):
+        """Track wires of the PauliSentence."""
+        return set().union(*(pw.wires for pw in self.keys()))
+
+    def to_mat(self, wire_order, format="dense"):
+        """Given a Pauli word, get the matrix representation.
```
Comment (albi3ro) [inline pennylane/pauli/pauli_arithmetic.py:239]:
```suggestion
        """Returns the matrix representation.
```

File: pennylane/pauli/pauli_arithmetic.py [hunk 29]
```diff
@@ -0,0 +1,280 @@
...
+
+    @property
+    def wires(self):
+        """Track wires in a PauliWord."""
+        return set(self)
+
+    def to_mat(self, wire_order, format="dense"):
+        """Given a Pauli word, get the matrix representation.
+
+        KeywordArgs:
```
Comment (Jaybsoni) [inline pennylane/pauli/pauli_arithmetic.py:161]:
```suggestion
        Keyword Args:
```

File: pennylane/pauli/pauli_arithmetic.py [hunk 30]
```diff
@@ -0,0 +1,280 @@
...
+        """Terminal representation for PauliWord"""
+        return str(self)
+
+    @property
+    def wires(self):
+        """Track wires in a PauliWord."""
+        return set(self)
+
+    def to_mat(self, wire_order, format="dense"):
+        """Given a Pauli word, get the matrix representation.
```
Comment (Jaybsoni) [inline pennylane/pauli/pauli_arithmetic.py:159]:
```suggestion
        """Returns the matrix representation.
```

**Review by albi3ro — APPROVED** [review]
Looks good to me! Exciting to finally have this in! 🚀

**Review by AlbertMitjans — APPROVED** [review]
Approved! 🚀
